dynamiq/memory/backends/dynamo_db.py

```diff
@@ -0,0 +1,421 @@
+            metadata_to_store[self.sort_key_name] = timestamp_decimal
+            metadata_to_store[self.partition_key_name] = message_id
+            processed_metadata = _convert_floats_to_decimals(metadata_to_store)
+            logger.debug(
+                f"Saving item with PK: {message_id}, SK: {timestamp_decimal},"
+                f" Metadata: {json.dumps(processed_metadata, default=str)}"
+            )
+            item = {
+                self.partition_key_name: message_id,
+                self.sort_key_name: timestamp_decimal,
+                self.role_attribute_name: message.role.value,
+                self.content_attribute_name: message.content,
+                self.metadata_attribute_name: processed_metadata,
+            }
+            self._dynamodb_table.put_item(Item=item)
+        except (ClientError, ValueError, TypeError) as e:
+            logger.error(f"Error adding message to DynamoDB: {e}")
+            raise DynamoDBMemoryError(f"Error adding message to DynamoDB: {e}") from e
+        except Exception as e:
+            logger.error(f"Unexpected error adding message to DynamoDB: {e}")
+            try:
+                logger.error(f"Problematic metadata (attempted): {json.dumps(message.metadata, default=str)}")
+            except Exception:
+                logger.error("Could not serialize problematic metadata for logging.")
+            raise DynamoDBMemoryError(f"Unexpected error adding message: {e}") from e
+
+    def _scan_and_process(
+        self, limit: int | None = None, filters: dict | None = None, query: str | None = None
+    ) -> list[Message]:
+        """Scans table, applies filters/query client-side, sorts, and limits."""
+        if self._dynamodb_table is None:
+            raise DynamoDBMemoryError("DynamoDB table not initialized.")
+        logger.warning(
+            f"Performing DynamoDB Scan on table '{self.index_name}' for search/get_all. This can be slow and costly."
+        )
+
+        scan_params = {}
+        processed_filters = _convert_floats_to_decimals(filters) if filters else None
+        filter_expression_attr = self._build_filter_expression_attr(processed_filters)
+        if filter_expression_attr:
+            scan_params["FilterExpression"] = filter_expression_attr
+            logger.debug(f"Using server-side FilterExpression object: {filter_expression_attr}")
+        elif filters:
+            logger.debug("No filters provided or FilterExpression could not be built.")
+
+        all_items = []
+        last_evaluated_key = None
+        items_processed = 0
+        fetch_multiplier = 10
+        scan_fetch_target = (limit * fetch_multiplier) if limit else 5000
+
+        try:
+            while True:
+                if last_evaluated_key:
+                    scan_params["ExclusiveStartKey"] = last_evaluated_key
+
+                scan_params["Limit"] = min(1000, scan_fetch_target - items_processed) if limit else 1000
```

Comment:
This magic number like 1000 and 5000 should be moved to constants (or maybe params) with clear naming